build-system/tasks/make-extension/index.js

```diff
@@ -24,304 +24,183 @@ const {
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
+ * Generate a sequence of all files in a directory recursively.
+ * @param {string} dir
+ * @yields {string}
  */
-
-import '../${name}';
-import {createElementWithAttributes} from '../../../../src/dom';
-
-describes.realWin(
-  '${name}',
-  {
-    amp: {
-      runtimeOn: true,
-      extensions: ['${name}'],
-    },
-  },
-  (env) => {
-    let win;
-    let element;
-
-    beforeEach(() => {
-      win = env.win;
-      element = createElementWithAttributes(win.document, '${name}', {
-        layout: 'responsive',
-      });
-      win.document.body.appendChild(element);
-    });
-
-    it('should contain "hello world" when built', async () => {
-      await element.whenBuilt();
-      expect(element.querySelector('div').textContent).to.equal('hello world');
-    });
+async function* walkDir(dir) {
+  for (const f of await fs.readdir(dir)) {
+    const dirPath = path.join(dir, f);
+
+    if ((await fs.stat(dirPath)).isDirectory()) {
+      yield* walkDir(dirPath);
+    } else {
+      yield path.join(dir, f);
+    }
   }
-);
-`;
 }
 
+const getTemplateDir = (template) => path.join(__dirname, 'template', template);
+
 /**
- * @param {string} name
- * @return {string}
- */
-function getJsExtensionFile(name) {
-  const className = pascalCase(name);
-  return `/**
- * Copyright ${year} The AMP HTML Authors. All Rights Reserved.
- *
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
- *
- *      http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS-IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
+ * @param {string} templateDir
+ * @param {Object<string, string>} replacements
+ * @param {string=} destinationDir
+ * @return {Array<string>}
  */
+async function writeFromTemplateDir(
+  templateDir,
+  replacements,
+  destinationDir = '.'
+) {
+  const destinationPath = (templatePath) =>
+    path.join(
+      destinationDir,
+      replace(templatePath.substr(templateDir.length + 1), replacements)
+    );
 
-import {CSS} from '../../../build/${name}.css';
-import {Layout} from '../../../src/layout';
+  const written = [];
+  for await (const templatePath of walkDir(templateDir)) {
+    const destination = destinationPath(templatePath);
 
-export class ${className} extends AMP.BaseElement {
-  /** @param {!AmpElement} element */
-  constructor(element) {
-    super(element);
+    // Skip if the destination file already exists
+    try {
+      await fs.access(destination);
```

Comment:
It looks like [`fs.access`](https://nodejs.org/api/fs.html#fs_fs_access_path_mode_callback) is not recommended to check if a file exists immediately before writing to it - can we do any of the following instead?
- attempt `fs.open` instead as described in one of the examples at the link
- skip this step / overwrite by default without introducing the flag -- Since we have git, it would be straight forward to discard unwanted overwriting.